### ISSUE TITLE
[BUG] Optimize Similarity Computations in Duplicate Ticket Service

### ISSUE THREAD

**Issue by Arpita2919** [riteshbonthalakoti/HELPDESK.AI#438]
### Bug Description

# [PERFORMANCE] Optimize Similarity Computations in Duplicate Ticket Service

## Category

Performance

## Problem Description

In `backend/services/duplicate_service.py`, the `check_duplicate` algorithm computes cosine similarity using a Python loop over stored embeddings:

```python
for ticket_id, stored_emb, _ in self._tickets:
    score = util.cos_sim(query_embedding, stored_emb).item()
```

This performs similarity calculations one tensor at a time, causing unnecessary Python-level iteration overhead and blocking the event loop as the number of stored tickets increases.

As ticket history grows, duplicate detection becomes increasingly slow and inefficient.

---

## Why It Matters

The current implementation does not scale efficiently for production workloads with hundreds or thousands of support tickets.

This can lead to:

* Increased API response latency
* Slower duplicate ticket detection
* Higher CPU usage
* Poor user experience during ticket submission

---

## Proposed Solution

Refactor the duplicate detection logic to use vectorized cosine similarity computation instead of per-ticket iteration.

Store embeddings as a stacked tensor and compute similarity scores in a single batch operation using PyTorch tensor operations.

This will significantly improve performance and scalability.

---

## Files Involved

* `backend/services/duplicate_service.py`

---

## Step-by-Step Implementation

1. Refactor ticket embedding storage to maintain a stacked tensor of embeddings
2. Replace the Python similarity loop with a vectorized cosine similarity operation
3. Compute similarity scores for all tickets in a single tensor operation
4. Use the highest similarity score to identify duplicate candidates
5. Verify duplicate detection accuracy remains unchanged
6. Benchmark performance improvements with larger ticket datasets

---

## Expected Outcome

* Faster duplicate ticket detection
* Reduced API latency
* Improved scalability for large ticket datasets
* More efficient embedding similarity computation

### Steps to Reproduce

1. Start the backend server with multiple stored ticket embeddings.
2. Submit a new support ticket for duplicate analysis.
3. Trigger the `check_duplicate` method in `backend/services/duplicate_service.py`.
4. Observe that cosine similarity is computed inside a Python loop for each stored embedding.
5. Notice increased response latency as the number of stored tickets grows.

### Expected Behavior

Duplicate ticket similarity checks should execute efficiently using vectorized tensor operations instead of per-ticket Python loops. API response times should remain stable even with large ticket datasets.

### Helpdesk.AI Version

main branch

**Comment by Arpita2919** [comment]
hey @ritesh-1918  i have raised issue #438. please assign it to me. Thank You!

**Comment by riteshbonthalakoti** [comment]
Hey @Arpita2919! 🙌

Thank you so much for your interest and commitment to tackling this open task: '[BUG] Optimize Similarity Computations in Duplicate Ticket Service'! This is a highly valued contribution to the platform.


### 🚀 Let's Connect & Elevate our Engineering Network!
As we build the future of automated IT operations with **HELPDESK.AI**, I would love to follow your development journey and keep you in the loop for future production-grade opportunities. Please take a quick moment to support the project and connect with me directly:

1. ⭐ **Star this repository**: Show your support and help HELPDESK.AI grow! [Star the repo here](https://github.com/ritesh-1918/HELPDESK.AI)
2. 🍴 **Fork this repository**: Create your own sandbox copy to experiment with our ML/OCR models! [Fork here](https://github.com/ritesh-1918/HELPDESK.AI/fork)
3. 👤 **Follow @ritesh-1918 on GitHub**: Stay updated on real-time open-source project releases! [Follow ritesh-1918](https://github.com/ritesh-1918)
4. 💼 **Connect on LinkedIn**: Let's build a strong, professional engineering connection for future collaborations! [Connect with Ritesh on LinkedIn](https://www.linkedin.com/in/ritesh1908/)

---

### 🔑 Mandatory Registration Steps
Before you raise your PR (strictly targeting the `gssoc` branch, NOT `main`), let's get you set up to test your code live on our deployed server:
- Head over to the live application: https://helpdeskaiv1.vercel.app/
- Create your standard user account at `https://helpdeskaiv1.vercel.app/signup`
- Reply to your assigned thread or email ritesh at `bonthalamadhavi1@gmail.com` with your username to get your access approved instantly!

**Comment by Exodus2004** [comment]
I have submitted a fix in PR #612. 

### Why this approach is superior:
1. **Parallel Vectorization:** Instead of iterating through each ticket embedding using a Python or loop—which triggers expensive interpreter overhead for every iteration—we stack all stored embeddings into a single PyTorch tensor using 	orch.stack and perform the cosine similarity calculation across the entire matrix at once.
2. **Reduced Latency:** By utilizing PyTorch's optimized internal matrix operations, we compute the similarity scores in parallel. This changes the computation from (N)$ sequential operations in Python land to a single parallelized tensor operation in C++ layer, scaling gracefully as the ticket history grows.
3. **Graceful Degradation:** The optimization preserves the existing fallback mechanisms so that if PyTorch or SentenceTransformers is unavailable (e.g., in developer/light mode), the application degrades cleanly instead of throwing unhandled exceptions.

**Comment by riteshbonthalakoti** [comment]
Superb suggestion, @Arpita2919! Triaged and yours to build.

While setting up, please ensure you've completed these 3 quick steps:
1. ⭐ **Star HELPDESK.AI**: https://github.com/ritesh-1918/HELPDESK.AI
2. 👤 **Follow the author**: https://github.com/ritesh-1918
3. 💼 **Connect on LinkedIn**: https://www.linkedin.com/in/ritesh1908/

Happy coding! Let's do it! 💻🔥

**Comment by tranha80** [comment]
### AI-Generated Solution Proposal

I've analyzed the issue and drafted a solution. Below is the proposed solution.

#### Solution Details:
```
```python
import torch
import torch.nn.functional as F

# --- Mocking sentence_transformers.util.cos_sim for demonstration ---
# In a real scenario, you'd import it or use F.cosine_similarity directly.
def mock_cos_sim(tensor1, tensor2):
    # If tensor1 is (1, dim) and tensor2 is (num_items, dim)
    # We want similarity between tensor1 and each row of tensor2
    return F.cosine_similarity(tensor1, tensor2, dim=-1)


# --- ORIGINAL IMPLEMENTATION (Conceptual, based on problem description) ---
# This class demonstrates the 'before' state for comparison.
class OriginalDuplicateService:
    def __init__(self):
        self._tickets = []  # List of (ticket_id, stored_embedding, other_data)
        self._next_id = 1

    def add_ticket(self, embedding: torch.Tensor):
        ticket_id = f"TICKET_{self._next_id}"
        self._tickets.append((ticket_id, embedding.squeeze(0), None)) # Store squeezed embedding
        self._next_id += 1
        return ticket_id

    def check_duplicate(self, query_embedding: torch.Tensor):
        # Ensure query_embedding is 1D for comparison with stored embeddings
        query_embedding = query_embedding.squeeze(0) 

        max_score = -1.0
        duplicate_id = None

        if not self._tickets:
            return None, max_score

        # PROBLEM: Iterates over each stored embedding
        for ticket_id, stored_emb, _ in self._tickets:
            # Assuming util.cos_sim takes two 1D tensors and returns a scalar tensor
            score = mock_cos_sim(query_embedding, stored_emb).item()
            if score > max_score:
                max_score = score
                duplicate_id = ticket_id
        return duplicate_id, max_score


# --- PROPOSED REFACTORED IMPLEMENTATION ---
# This class demonstrates the 'after' state with vectorized operations.
class OptimizedDuplicateService:
    def __init__(self):
        self._ticket_ids = []
        self._stored_embeddings = None  # Will be a stacked tensor (num_tickets, embedding_dim)
        self._next_id = 1

    def add_ticket(self, embedding: torch.Tensor):
        ticket_id = f"TICKET_{self._next_id}"
        self._ticket_ids.append(ticket_id)
        
        # Ensure embedding is (1, embedding_dim) for stacking
        embedding = embedding.unsqueeze(0) if embedding.dim() == 1 else embedding

        if self._stored_embeddings is None:
            self._stored_embeddings = embedding
        else:
            # Stack new embedding with existing ones along dimension 0
            self._stored_embeddings = torch.cat((self._stored_embeddings, embedding), dim=0)
            
        self._next_id += 1
        return ticket_id

    def check_duplicate(self, query_embedding: torch.Tensor):
        if self._stored_embeddings is None or not self._ticket_ids:
            return None, -1.0

        # Ensure query_embedding is (1, embedding_dim) for vectorized comparison
        query_embedding = query_embedding.unsqueeze(0) if query_embedding.dim() == 1 else query_embedding

        # SOLUTION: Vectorized cosine similarity computation
        # F.cosine_similarity handles broadcasting when one tensor has a single item (query_embedding)
        # It will compute similarity between query_embedding and each row of _stored_embeddings
        # Resulting 'scores' will be a 1D tensor of shape (num_tickets,)
        scores = F.cosine_similarity(query_embedding, self._stored_embeddings, dim=1)

        # Find the maximum score and its index
        max_score, max_index = torch.max(scores, dim=0)
        
        # Map the index back to the original ticket ID
        duplicate_id = self._ticket_ids[max_index.item()]

        return duplicate_id, max_score.item()


# --- DEMONSTRATION & VERIFICATION ---
if __name__ == "__main__":
    # Generate some dummy embeddings
    embedding_dim = 768
    num_existing_tickets = 1000 # Simulating a growing history
    
    # Generate random embeddings. In a real scenario, these would come from your model.
    existing_embeddings = [torch.randn(1, embedding_dim) for _ in range(num_existing_tickets)]
    
    # A query embedding that might be a duplicate of one of the existing ones
    query_embedding_true_duplicate = existing_embeddings[num_existing_tickets // 2] + torch.randn(1, embedding_dim) * 0.05 # Slightly perturbed version
    query_embedding_new = torch.randn(1, embedding_dim) # A completely new one

    print("--- Original Service Test ---")
    original_service = OriginalDuplicateService()
    for emb in existing_embeddings:
        original_service.add_ticket(emb)
    
    # Test with a known duplicate-like query
    original_dup_id, original_dup_score = original_service.check_duplicate(query_embedding_true_duplicate)
    print(f"Original Service: Found duplicate '{original_dup_id}' with score {original_dup_score:.4f}")
    
    # Test with a new query
    original_new_id, original_new_score = original_service.check_duplicate(query_embedding_new)
    print(f"Original Service: Found duplicate '{original_new_id}' with score {original_new_score:.4f} (expected low for new)")
    

    print("\n--- Optimized Service Test ---")
    optimized_service = OptimizedDuplicateService()
    for emb in existing_embeddings:
        optimized_service.add_ticket(emb)

    # Test with a known duplicate-like query
    optimized_dup_id, optimized_dup_score = optimized_service.check_duplicate(query_embedding_true_duplicate)
    print(f"Optimized Service: Found duplicate '{optimized_dup_id}' with score {optimized_dup_score:.4f}")
    
    # Test with a new query
    optimized_new_id, optimized_new_score = optimized_service.check_duplicate(query_embedding_new)
    print(f"Optimized Service: Found duplicate '{optimized_new_id}' with score {optimized_new_score:.4f} (expected low for new)")

    print("\n--- Verification ---")
    # Assert that the logic yields the same results (within float precision)
    assert original_dup_id == optimized_dup_id, "Duplicate IDs don't match!"
    assert torch.isclose(torch.tensor(original_dup_score), torch.tensor(optimized_dup_score), atol=1e-4), "Duplicate scores don't match!"
    assert original_new_id == optimized_new_id, "New IDs don't match!"
    assert torch.isclose(torch.tensor(original_new_score), torch.tensor(optimized_new_score), atol=1e-4), "New scores don't match!"
    print("Both services yield consistent duplicate detection results.")

    # --- Performance Hint ---
    # To truly benchmark, one would use timeit or the 'perf_counter' module 
    # and run check_duplicate many times for both services with a large dataset.
    # For instance:
    # import time
    # start_time = time.perf_counter()
    # # ... call check_duplicate many times for original_service ...
    # end_time = time.perf_counter()
    # print(f"Original service time: {end_time - start_time}")
    # # Repeat for optimized_service
    # This will clearly show the order-of-magnitude performance improvement.
```
```

#### Proposal Letter:
Dear [Hiring Manager Name/Client Name],

I am writing to express my keen interest in the freelance opportunity to address the performance bug outlined in your job description: "[BUG] Optimize Similarity Computations in Duplicate Ticket Service."

As an experienced freelancer specializing in performance optimization and scalable backend solutions, I immediately recognized the critical bottleneck described in `backend/services/duplicate_service.py`. The current method of iterating through individual embeddings and computing cosine similarity one by one is a classic anti-pattern for scalability in modern deep learning applications. It inevitably leads to the high latency, increased CPU usage, and degraded user experience you've observed as your ticket history grows.

Your analysis and proposed solution are spot-on. Refactoring to leverage vectorized cosine similarity computation with a stacked tensor of embeddings is precisely the robust and efficient approach needed. I have extensive experience implementing such optimizations using PyTorch (or similar libraries) to transform slow, iterative processes into lightning-fast batch operations.

My proposed approach will strictly adhere to the steps you've outlined, ensuring a comprehensive and verified solution:

1.  **Refactor Embedding Storage:** I will modify the internal storage mechanism within `DuplicateService` to maintain a single, stacked `torch.Tensor` containing all stored embeddings, alongside a separate, efficient mapping for ticket IDs.
2.  **Vectorized Similarity Computation:** The Python loop for `check_duplicate` will be replaced with a single, highly optimized tensor operation (e.g., using `torch.matmul` combined with appropriate normalization) to compute similarity scores for the query embedding against all stored embeddings simultaneously.
3.  **Efficient Duplicate Identification:** I will then efficiently extract the highest similarity score and its corresponding ticket ID from the resulting score tensor.
4.  **Accuracy Verification:** Post-implementation, I will rigorously test to ensure that the duplicate detection accuracy remains identical to the original logic, preventing any regressions.
5.  **Performance Benchmarking:** Crucially, I will implement performance benchmarks using synthetic larger ticket datasets to quantify the improvements in detection speed and confirm the expected reduction in latency and CPU overhead.

I am confident that I can deliver a significant performance boost, making your duplicate ticket service highly scalable and responsive. My commitment extends beyond just writing code; I prioritize clean, maintainable code, thorough testing, and clear documentation.

I am available to start immediately and anticipate completing this task within [mention a realistic timeframe, e.g., 2-3 business days] days, including testing and benchmarking, depending on the complexity of the existing codebase beyond the described snippet. I am keen to discuss this project further and provide a more precise timeline and cost estimate.

Thank you for considering my application. I look forward to the opportunity to optimize your service.

Best regards,

[Your Name/Freelancer Name]
[Your Professional Title]
[Your Contact Information]

---
**Payment Link (USD):** [Pay via USD](https://www.sandbox.paypal.com/checkoutnow?token=fallback_mock_508640)
*Please verify the solution and approve payment.*

**Comment by copernicusjones** [comment]
Interested in tackling this performance optimization bounty for HELPDESK.AI.

**My approach:**

The current loop-based cosine similarity is O(n) Python-level iteration — exactly the kind of thing that should be vectorized. Here's my plan:

1. **Refactor storage**: Replace the list-of-tuples `_tickets` structure with a single stacked `torch.Tensor` of shape `(N, embedding_dim)`, plus a parallel list of metadata (ticket IDs, timestamps)
2. **Vectorized similarity**: Replace the for-loop with a single `torch.nn.functional.cosine_similarity(query.unsqueeze(0), stored_matrix, dim=1)` call → returns all scores in one GPU/CPU tensor op
3. **Batch-friendly**: For multiple query support, `torch.mm(query_batch, stored_matrix.T)` gives you the full similarity matrix in one matmul
4. **Benchmark**: I'll add a benchmark script showing before/after timing at 100, 1K, 10K, 100K ticket scale

This should give you **100-1000x speedup** at scale since you're replacing Python-level iteration with a single BLAS-backed tensor operation.

**Stack**: PyTorch, pytest for validation that accuracy is unchanged before/after.

GitHub: [@copernicusjones](https://github.com/copernicusjones)

**Comment by rajesh-puripanda** [comment]
I'll optimize the duplicate ticket similarity computations by caching embedding vectors, using approximate nearest neighbor search instead of brute-force, and batch-processing comparisons. Benchmarks included. I would like to be assigned this issue and will start working on it.

**Comment by riteshbonthalakoti** [comment]
Hi @Arpita2919 @Exodus2004 @copernicusjones @rajesh-puripanda @tranha80! 🙌

Thank you so much for your active interest and engagement with this open task: **"[BUG] Optimize Similarity Computations in Duplicate Ticket Service"**! 🚀

Please complete the **mandatory onboarding steps** below to qualify for points on the GSSoC leaderboard. Remember, **following the admin's account manually** is a required step before any PRs can be merged!

Ritesh is actively reviewing and will be merging PRs very soon. Let's build something amazing! 💻🔥


---

### 🌟 Mandatory Onboarding Steps (Leaderboard Eligibility)
Before you raise your PR (strictly targeting the `gssoc` branch, NOT `main`), you **MUST** complete these required steps to finalize your points:
1. ⭐ **Star this repository**: https://github.com/ritesh-1918/HELPDESK.AI (Mandatory)
2. 👤 **Follow the Project Admin**: https://github.com/ritesh-1918 (Mandatory - manual step)
3. 💼 **Connect on LinkedIn**: https://www.linkedin.com/in/ritesh1908/ (Mandatory)

*Note: Access to the live testing dashboard will only be approved for developers who have completed these steps.*

**Comment by riteshbonthalakoti** [comment]
Hi contributors! 🙌

This issue has been successfully resolved and integrated into the `gssoc` branch via our merged pull requests! 

Thank you all for the outstanding code and contributions to the project. This issue is officially closed as completed! 🚀💻
